### PR TITLE
Feature/feedback disambiguation

### PR DESCRIPTION
**Pull request by CDRussell**
Task/Issue URL: https://app.asana.com/0/715106103902962/1109166030178654
iOS equivalent PR: https://github.com/duckduckgo/iOS/pull/420
Tech Design URL: 
CC

# Description
Adds a new flow for users to leave feedback from the `Settings` page.

This new flow tries to better capture the nature of the feedback by getting users to identify the category of the problem (is their feedback about missing features, bugs, search results etc...)

## Description of the flow

There is an initial happy / sad clarifier.
Screenshot 2019-03-22 at 17 07 34

#### Rate, give more details or finish
In the happy flow, the users will be offered to chance to give us more info or to rate us, or to end the journey there.

They will only see this screen if they can rate us (app installed from Play Store) - otherwise this screen will be skipped and they'll be shown the "more details" page instead.

device-2019-03-22-172037

#### More details page
Screenshot 2019-03-22 at 17 07 38

### Sad Flow

In this flow, the user will be shown a list of main reasons why they might be unhappy and asked to pick from the list.
Screenshot 2019-03-22 at 17 07 46

Depending on the option chosen, they'll either navigate to choosing a subreason, a page to provide "extra details", or they'll see a broken site page.

When choosing a subreason, they'll always end with navigating to the open ended "share details" page.

#### Missing Browser features
This will show subcategories. Choosing a subcategory will navigate to the open ended "share details" page.

#### Website loading issues
This will show a new broken site page (note, the existing "report broken site" screen is left as is)

#### Search results not good enough
This will show show subcategories. Choosing a subcategory will navigate to the open ended "share details" page.

#### Customization 
This will show subcategories. Choosing a subcategory will navigate to the open ended "share details" page.

#### None of these
This will jump straight to the open ended "share details" page.

**Steps to test this PR**:
1. Test all the flows, for each path observing pixel data and and the HTTP POST network request that happens when (if) you reach the terminal journey in that flow
1. Repeat with other (dark/light) theme

See the internal ticket for where details on where to look to verify pixel and POST data.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: app/src/main/java/com/duckduckgo/app/settings/SettingsActivity.kt [hunk 1]
```diff
@@ -193,6 +210,7 @@ class SettingsActivity : DuckDuckGoActivity(), SettingsAutomaticallyClearWhatFra
     companion object {
         private const val CLEAR_WHAT_DIALOG_TAG = "CLEAR_WHAT_DIALOG_FRAGMENT"
         private const val CLEAR_WHEN_DIALOG_TAG = "CLEAR_WHEN_DIALOG_FRAGMENT"
+        private const val FEEDBACK_RC = 100
```
Comment (subsymbolic) [inline app/src/main/java/com/duckduckgo/app/settings/SettingsActivity.kt]: `RC` doesn't seem clear to me. Can we expand it to `FEEDBACK_RESPONSE_CODE`?

File: app/src/main/java/com/duckduckgo/app/feedback/ui/negative/subreason/SubReasonNegativeFeedbackFragment.kt [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    private fun moreCustomizationsSubReasons(): List<FeedbackTypeSubReasonDisplay> {
+        return FeedbackType.CustomizationSubReasons.values().mapNotNull {
+            FeedbackTypeDisplay.subReasons[it]
+        }
+    }
+
+    private fun appSlowOrBuggySubReasons(): List<FeedbackTypeSubReasonDisplay> {
+        return FeedbackType.PerformanceSubReasons.values().mapNotNull() {
```
Comment (subsymbolic) [inline app/src/main/java/com/duckduckgo/app/feedback/ui/negative/subreason/SubReasonNegativeFeedbackFragment.kt]: No need for brackets on `mapNotNull`

File: app/src/main/java/com/duckduckgo/app/feedback/ui/negative/FeedbackTypeDisplay.kt [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+import com.duckduckgo.app.feedback.ui.negative.FeedbackType.*
+import com.duckduckgo.app.feedback.ui.negative.FeedbackType.CustomizationSubReasons.*
+import com.duckduckgo.app.feedback.ui.negative.FeedbackType.MainReason.*
+import com.duckduckgo.app.feedback.ui.negative.FeedbackType.MainReason.OTHER
+import com.duckduckgo.app.feedback.ui.negative.FeedbackType.MissingBrowserFeaturesSubReasons.*
+import com.duckduckgo.app.feedback.ui.negative.FeedbackType.PerformanceSubReasons.*
+import com.duckduckgo.app.feedback.ui.negative.FeedbackType.SearchNotGoodEnoughSubReasons.*
+
+fun MainReason.displayText(): FeedbackTypeDisplay.FeedbackTypeMainReasonDisplay? = FeedbackTypeDisplay.mainReasons[this]
+fun SubReason.displayText(): FeedbackTypeDisplay.FeedbackTypeSubReasonDisplay? = FeedbackTypeDisplay.subReasons[this]
```
Comment (subsymbolic) [inline app/src/main/java/com/duckduckgo/app/feedback/ui/negative/FeedbackTypeDisplay.kt]: `SubReason.displayText()` looks unused in IDE.

File: app/src/main/res/layout/content_feedback.xml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        <ImageView
+            android:id="@+id/positiveFeedbackButton"
+            android:layout_width="100dp"
+            android:layout_height="100dp"
+            android:layout_marginTop="34dp"
+            android:background="?attr/selectableItemBackgroundBorderless"
+            android:clickable="true"
+            android:contentDescription="@string/feedbackInitialDisambiguationHappyButtonContentDescription"
+            android:focusable="true"
+            tools:src="@drawable/ic_button_happy_light_theme"
```
Comment (subsymbolic) [inline app/src/main/res/layout/content_feedback.xml]: This drawable doesn't exist.

File: app/src/main/res/layout/content_feedback.xml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toBottomOf="@id/subtitle" />
+
+        <ImageView
+            android:id="@+id/negativeFeedbackButton"
+            android:layout_width="100dp"
+            android:layout_height="100dp"
+            android:background="?attr/selectableItemBackgroundBorderless"
+            android:contentDescription="@string/feedbackInitialDisambiguationSadButtonContentDescription"
+            tools:src="@drawable/ic_button_sad_light_theme"
```
Comment (subsymbolic) [inline app/src/main/res/layout/content_feedback.xml]: This drawable doesn't exist.

**Comment by subsymbolic** [comment]
@CDRussell I have reviewed and tested the code and validated that the flows make sense and work. In terms of verifying every permutation, I haven't done that. Is this something Azad/Chris could do?

File: app/src/main/java/com/duckduckgo/app/brokensite/BrokenSiteActivity.kt [hunk 6]
```diff
@@ -92,12 +92,12 @@ class FeedbackActivity : DuckDuckGoActivity() {
...
 
     private fun confirmAndFinish() {
-        longToast(R.string.feedbackSubmitted)
+        longToast(R.string.brokenSiteSubmitted)
         finishAfterTransition()
     }
 
     private fun render(viewState: ViewState) {
-        val messageHint = if (viewState.isBrokenSite) R.string.feedbackBrokenSiteHint else R.string.feedbackMessageHint
+        val messageHint = if (viewState.isBrokenSite) R.string.brokenSiteBrokenSiteHint else R.string.brokenSiteMessageHint
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/brokensite/BrokenSiteActivity.kt]: Good feedback; the legacy broken site / feedback dialog is now solely used for providing broken site reports; no option of providing general feedback through it.

File: app/src/main/java/com/duckduckgo/app/survey/api/SurveyGroup.kt [hunk 7]
```diff
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-package com.duckduckgo.app.feedback.api
+package com.duckduckgo.app.survey.api
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/survey/api/SurveyGroup.kt]: Good spot; i've opted to move them all to `survey` as `feedback` is crowded as is.

File: app/src/main/java/com/duckduckgo/app/statistics/pixels/Pixel.kt [hunk 8]
```diff
@@ -138,7 +141,7 @@ class ApiBasedPixel @Inject constructor(
             .subscribe({
                 Timber.v("Pixel sent: $pixelName")
             }, {
-                Timber.w("Pixel failed: $pixelName", it)
+                Timber.w(it, "Pixel failed: $pixelName")
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/statistics/pixels/Pixel.kt]: Fair enough; me too. I just noticed that `Timber.w("Pixel failed: $pixelName", it)` is clearly not what anyone intended.

But let's just remove the `it` throwable altogether from the logging.

File: app/src/main/java/com/duckduckgo/app/settings/SettingsActivity.kt (hunk 1 above)
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/settings/SettingsActivity.kt]: Expanded to `FEEDBACK_REQUEST_CODE`. That `RC` was ambiguous between request and response is further evidence I should have done that all along! 😄

File: app/src/androidTest/java/com/duckduckgo/app/feedback/ui/common/FeedbackViewModelTest.kt [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+    fun whenUserProvidesFurtherDetailsForPositiveFeedbackThenFeedbackSubmitted() = runBlocking<Unit> {
+        withContext(Dispatchers.Main) {
+            testee.userProvidedPositiveOpenEndedFeedback("foo")
+        }
+
+        verify(feedbackSubmitter).sendPositiveFeedback("foo")
+    }
+
+    @Test
+    fun whenUserProvidesFurtherDetailsForPositiveFeedbackThenExitCommandIssued() = runBlocking<Unit> {
```
Comment (subsymbolic) [inline app/src/androidTest/java/com/duckduckgo/app/feedback/ui/common/FeedbackViewModelTest.kt]: This test is intermittently failing, see https://app.bitrise.io/build/68a947da5eaa70d9.

Comment (CDRussell) [reply app/src/androidTest/java/com/duckduckgo/app/feedback/ui/common/FeedbackViewModelTest.kt]: Made a change; I've ran the CI quite a few times in the hopes that it's now stable.

**Pull request merged**